File: scripts/enrich-skills.py
#!/usr/bin/env python3
"""
enrich-skills.py
Fetches GitHub URLs from clawhub inspect metadata for skills in data/skills.json.
Run: python3 scripts/enrich-skills.py
"""
import subprocess
import json
import re
import time
import sys

CLAWHUB = '/home/sean/.nvm/versions/node/v22.22.1/bin/clawhub'

def fetch_github_url(slug):
    """Run clawhub inspect --file SKILL.md and extract repository/homepage from metadata."""
    try:
        result = subprocess.run(
            [CLAWHUB, 'inspect', slug, '--file', 'SKILL.md'],
            capture_output=True,
            text=True,
            timeout=20
        )
        output = result.stdout + result.stderr

        # Output structure: --- [metadata] --- [SKILL.md content]
        # Split by --- separator; metadata is in parts[1]
        parts = output.split('---')
        if len(parts) < 2:
            return None

        metadata = parts[1]  # Metadata section is after first ---

        # Look for repository or homepage fields
        repo_match = re.search(r'repository:\s*(https?://github\.com/[^\s]+)', metadata, re.IGNORECASE)
        if repo_match:
            url = repo_match.group(1).rstrip('/')
            if url.endswith('.git'):
                url = url[:-4]
            return url

        home_match = re.search(r'homepage:\s*(https?://github\.com/[^\s]+)', metadata, re.IGNORECASE)
        if home_match:
            url = home_match.group(1).rstrip('/')
            if url.endswith('.git'):
                url = url[:-4]
            return url

        return None
    except Exception:
        return None

def main():
    with open('data/skills.json', 'r') as f:
        skills = json.load(f)

    print('Loaded {} skills'.format(len(skills)))

    # Skills that need GitHub URL
    need_gh = [s for s in skills if not s.get('githubUrl')]
    print('Skills needing GitHub URL: {}'.format(len(need_gh)))

    updated = 0
    errors = 0

    for i, skill in enumerate(need_gh):
        slug = skill['slug']

        # Check for "Skill not found" to skip removed skills
        if (i + 1) % 10 == 0:
            # Batch delay every 10 skills
            time.sleep(1.5)

        url = fetch_github_url(slug)

        if url:
            skill['githubUrl'] = url
            updated += 1
            print('[{}] {} => {}'.format(i+1, slug, url))
        else:
            # Check if skill was not found
            try:
                result = subprocess.run(
                    [CLAWHUB, 'inspect', slug],
                    capture_output=True, text=True, timeout=10
                )
                out = result.stdout + result.stderr
                if 'not found' in out.lower() or 'error' in out.lower():
                    print('[{}] {} => NOT FOUND (skipping)'.format(i+1, slug))
                else:
                    print('[{}] {} => (no GH repo)'.format(i+1, slug))
            except Exception:
                print('[{}] {} => ERROR'.format(i+1, slug))
                errors += 1

    print('\nUpdated {} GitHub URLs'.format(updated))
    if errors:
        print('Errors: {}'.format(errors))

    # Re-sort by score
    skills.sort(key=lambda s: s['score'], reverse=True)
    for i, s in enumerate(skills):
        s['rank'] = i + 1

    with open('data/skills.json', 'w') as f:
        json.dump(skills, f, indent=2, ensure_ascii=False)

    print('Saved data/skills.json ({} skills)'.format(len(skills)))

if __name__ == '__main__':
    main()

File: scripts/enrich-skills.js
/**
 * enrich-skills.js
 * Fetches stars and GitHub URL for each skill in data/skills.json
 * using concurrent API calls. Updates data/skills.json in place.
 *
 * GitHub URL is extracted from SKILL.md file content (clawhub inspect --file SKILL.md)
 * since the API metadata field does not contain repository info.
 */
const { execSync } = require('child_process');
const fs = require('fs');
const https = require('https');

const CLAWHUB = '/home/sean/.nvm/versions/node/v22.22.1/bin/clawhub';

const skills = JSON.parse(fs.readFileSync('data/skills.json', 'utf8'));
console.log(`Enriching ${skills.length} skills...`);

// Skills already enriched
const doneStars = skills.filter(s => s.stars !== undefined).length;
const doneGithub = skills.filter(s => s.githubUrl).length;
console.log(`Already have stars: ${doneStars}/${skills.length}`);
console.log(`Already have githubUrl: ${doneGithub}/${skills.length}`);

// Fetch stars via HTTPS API
function fetchSkillStats(slug) {
  return new Promise((resolve) => {
    const url = `https://clawhub.ai/api/v1/skills/${slug}`;
    https.get(url, { headers: { 'User-Agent': 'clawhub-client/1.0' } }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const json = JSON.parse(data);
          const stats = json?.skill?.stats;
          resolve({ slug, stars: stats?.stars ?? null });
        } catch {
          resolve({ slug, stars: null });
        }
      });
    }).on('error', () => resolve({ slug, stars: null }));
  });
}

// Extract GitHub URL from SKILL.md content using clawhub inspect
function fetchGithubUrl(slug) {
  return new Promise((resolve) => {
    try {
      // clawhub inspect --file SKILL.md outputs:
      // ---  (separator line)
      // [file content]
      // ---  (separator)
      // Skill metadata...
      const out = execSync(
        `${CLAWHUB} inspect ${slug} --file SKILL.md 2>&1`,
        { encoding: 'utf8', timeout: 15000 }
      );

      // Extract GitHub URL from SKILL.md content (between --- separators)
      const parts = out.split('---');
      let content = '';
      if (parts.length >= 2) {
        // Content is between first two --- markers
        content = parts.slice(1, -1).join('---').trim();
      } else {
        content = out;
      }

      // Look for GitHub repository URLs in the content
      const githubPatterns = [
        /(?:github|repository|source)\s*[:=]\s*(https?:\/\/github\.com\/[^\s)\]"'>]+)/gi,
        /(https?:\/\/github\.com\/[a-zA-Z0-9_.-]+\/[a-zA-Z0-9_.-]+(?:\.git)?)/gi,
        /git\s+(https?:\/\/github\.com\/[^\s]+)/gi,
      ];

      for (const pattern of githubPatterns) {
        pattern.lastIndex = 0;
        const match = pattern.exec(content);
        if (match) {
          let url = match[1] || match[0];
          // Remove trailing .git if present
          url = url.replace(/\.git$/, '');
          resolve({ slug, githubUrl: url });
          return;
        }
      }
      resolve({ slug, githubUrl: null });
    } catch (e) {
      resolve({ slug, githubUrl: null });
    }
  });
}

async function run() {
  // Step 1: Fetch stars for skills missing them (batch size 20, concurrent)
  const starsBatchSize = 20;
  let processed = 0;

  for (let i = 0; i < skills.length; i += starsBatchSize) {
    const batch = skills.slice(i, i + starsBatchSize);
    const needsFetch = batch.filter(s => s.stars === undefined);

    if (needsFetch.length > 0) {
      const results = await Promise.all(needsFetch.map(s => fetchSkillStats(s.slug)));
      const resultMap = new Map(results.map(r => [r.slug, r]));
      for (const skill of batch) {
        const result = resultMap.get(skill.slug);
        if (result && result.stars !== null) {
          skill.stars = result.stars;
        }
      }
    }

    processed += batch.length;
    process.stdout.write(`\rStars: ${processed}/${skills.length}     `);
    await new Promise(r => setTimeout(r, 200));
  }
  console.log('\nStars enrichment complete.');

  // Step 2: Fetch GitHub URLs for skills missing them
  // Use clawhub inspect --file SKILL.md to extract repository URLs
  const needsGithub = skills.filter(s => !s.githubUrl);
  console.log(`Fetching GitHub URLs for ${needsGithub.length} skills...`);

  const ghBatchSize = 5; // Small batch to avoid rate limiting
  let ghProcessed = 0;

  for (let i = 0; i < needsGithub.length; i += ghBatchSize) {
    const batch = needsGithub.slice(i, i + ghBatchSize);
    const results = await Promise.all(batch.map(s => fetchGithubUrl(s.slug)));
    const resultMap = new Map(results.map(r => [r.slug, r]));

    for (const skill of skills) {
      const result = resultMap.get(skill.slug);
      if (result && result.githubUrl) {
        skill.githubUrl = result.githubUrl;
      }
    }

    ghProcessed += batch.length;
    process.stdout.write(`\rGitHub: ${ghProcessed}/${needsGithub.length}     `);

    // Delay between batches to avoid rate limiting
    if (i + ghBatchSize < needsGithub.length) {
      await new Promise(r => setTimeout(r, 1000));
    }
  }
  console.log('\nGitHub URL enrichment complete.');

  // Sort by score and reassign ranks
  skills.sort((a, b) => b.score - a.score);
  skills.forEach((s, i) => s.rank = i + 1);

  fs.writeFileSync('data/skills.json', JSON.stringify(skills, null, 2));
  console.log('Updated data/skills.json');

  const finalGithub = skills.filter(s => s.githubUrl).length;
  console.log(`Final stats: ${skills.length} skills, ${finalGithub} with GitHub URLs`);
}

run().catch(console.error);
